File: Stack/225_ImplementStackusingQueues.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# @Author: xuezaigds@gmail.com
from collections import deque


class Stack(object):
    def __init__(self):
        self._queue = deque()

    def push(self, x):
        # Pushing to back and
        # then rotating the queue until the new element is at the front
        q = self._queue
        q.append(x)
        for i in xrange(len(q) - 1):
            q.append(q.popleft())

    def pop(self):
        self._queue.popleft()

    def top(self):
        return self._queue[0]

    def empty(self):
        return not len(self._queue)

"""Test
if __name__ == '__main__':
    s = Stack()
    s.push(1)
    s.push(2)
    print s.top()
    s.pop()
    print s.empty()
    print s.top()
    s.pop()
    print s.empty()
"""

File: Stack/225_ImplementStackusingQueues.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-24 13:24:37
 */

class Stack {
public:
    // Push element x onto stack.
    // then rotating the queue until the new element is at the front
    void push(int x) {
        q.push(x);
        for(int i=0;i<q.size()-1;i++){
            q.push(q.front());
            q.pop();
        }
    }

    // Removes the element on top of the stack.
    void pop() {
        q.pop();
    }

    // Get the top element.
    int top() {
        return q.front();
    }

    // Return whether the stack is empty.
    bool empty() {
        return q.empty();
    }
private:
    queue<int> q{};
};
